lib/web_ui/lib/src/engine/canvaskit/font_fallbacks.dart

```diff
@@ -0,0 +1,690 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.12
+part of engine;
+
+Future<void> _findFontsForMissingCodeunits(List<int> codeunits) async {
+  _ensureNotoFontTreeCreated();
+  Set<_NotoFont> fonts = <_NotoFont>{};
+  for (int codeunit in codeunits) {
+    fonts.addAll(_lookupNotoFontsForCodeunit(codeunit));
+  }
+  fonts = _findMinimumFontsForCodeunits(codeunits, fonts);
+  for (_NotoFont font in fonts) {
+    if (_resolvedNotoFonts[font] == null) {
+      String googleFontCss = await html.window
+          .fetch(font.googleFontsCssUrl)
+          .then((dynamic response) =>
+              response.text().then<String>((dynamic x) => x as String));
+      final _ResolvedNotoFont resolvedFont =
+          _makeResolvedNotoFontFromCss(googleFontCss, font.name);
+      _registerResolvedFont(font, resolvedFont);
+    }
+  }
+
+  Set<_ResolvedNotoSubset> resolvedFonts = <_ResolvedNotoSubset>{};
+  for (int codeunit in codeunits) {
+    resolvedFonts.addAll(_lookupResolvedFontsForCodeunit(codeunit));
+  }
+
+  for (_ResolvedNotoSubset resolvedFont in resolvedFonts) {
+    skiaFontCollection.registerFallbackFont(
+        resolvedFont.url, resolvedFont.name);
+  }
+  await skiaFontCollection.ensureFontsLoaded();
+  sendFontChangeMessage();
+}
+
+_ResolvedNotoFont _makeResolvedNotoFontFromCss(String css, String name) {
```

Comment:
Please add doc comment. Ideally paste some example part of the file here so its easier to read parsing code below.

Reply:
Done